File: service/app/src/test/java/io/miragon/blueprint/process/BikeLeasingProcessTest.java
package io.miragon.blueprint.process;

import io.miragon.blueprint.adapter.process.BikeLeasingProcessProcessApi.Elements;
import io.miragon.blueprint.adapter.process.CancelBikeOrderProcessApi;
import io.miragon.blueprint.application.port.inbound.ActivateLeasingUseCase;
import io.miragon.blueprint.application.port.inbound.BookCancellationCostsUseCase;
import io.miragon.blueprint.application.port.inbound.CancelContractUseCase;
import io.miragon.blueprint.application.port.inbound.CancelInsurancePolicyUseCase;
import io.miragon.blueprint.application.port.inbound.IssueInsurancePolicyUseCase;
import io.miragon.blueprint.application.port.inbound.OrderBikeUseCase;
import io.miragon.blueprint.application.port.inbound.RejectApplicationUseCase;
import io.miragon.blueprint.application.port.inbound.RequestOrderCancellationUseCase;
import io.miragon.blueprint.application.port.inbound.SendCancellationConfirmationUseCase;
import io.miragon.blueprint.application.port.inbound.SendContractUseCase;
import io.miragon.blueprint.application.port.inbound.SendSignatureReminderUseCase;
import io.miragon.blueprint.application.port.inbound.ValidateApplicationUseCase;
import io.miragon.blueprint.application.port.outbound.LeasingProcess;
import io.miragon.blueprint.domain.bike.BikeId;
import io.miragon.blueprint.domain.bike.OrderId;
import io.miragon.blueprint.domain.leasing.ApplicationId;
import io.miragon.blueprint.domain.leasing.CustomerName;
import io.miragon.blueprint.domain.leasing.Email;
import io.miragon.blueprint.domain.leasing.LeasingApplication;
import io.miragon.blueprint.domain.leasing.LeasingStatus;
import java.time.LocalDateTime;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.operaton.bpm.engine.ProcessEngine;
import org.operaton.bpm.engine.RuntimeService;
import org.operaton.bpm.engine.TaskService;
import org.operaton.bpm.engine.runtime.ProcessInstance;
import org.operaton.bpm.engine.task.Task;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.bean.override.mockito.MockitoBean;

import static io.miragon.blueprint.process.util.JobExecutionUtils.continueToNextWaitState;
import static io.miragon.blueprint.process.util.ProcessInstanceUtils.findProcessInstance;
import static io.miragon.blueprint.process.util.TimerUtils.fireTimer;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;
import static org.mockito.Mockito.atLeast;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.operaton.bpm.engine.test.assertions.bpmn.BpmnAwareTests.assertThat;
import static org.operaton.bpm.engine.test.assertions.bpmn.BpmnAwareTests.init;

/**
 * Drives the deployed model end-to-end. Unlike the classic-delegate blueprint, the service tasks are
 * external tasks completed by the real, asynchronously polling {@code @ProcessEngineWorker} beans (which
 * call the mocked use cases). Each step therefore drives the process with
 * {@link io.miragon.blueprint.process.util.JobExecutionUtils#continueToNextWaitState continueToNextWaitState},
 * which fires async continuations <em>and</em> waits for the workers to drain the open external tasks, then
 * correlates messages / fires timers / completes the user task through the {@link LeasingProcess} port.
 */
@SpringBootTest
@ActiveProfiles("test")
public class BikeLeasingProcessTest {

    @Autowired
    private LeasingProcess process;

    @Autowired
    private RuntimeService runtimeService;

    @Autowired
    private TaskService taskService;

    @Autowired
    private ProcessEngine processEngine;

    @MockitoBean
    private ValidateApplicationUseCase validateApplicationUseCase;

    @MockitoBean
    private RejectApplicationUseCase rejectApplicationUseCase;

    @MockitoBean
    private SendContractUseCase sendContractUseCase;

    @MockitoBean
    private CancelContractUseCase cancelContractUseCase;

    @MockitoBean
    private IssueInsurancePolicyUseCase issueInsurancePolicyUseCase;

    @MockitoBean
    private CancelInsurancePolicyUseCase cancelInsurancePolicyUseCase;

    @MockitoBean
    private SendSignatureReminderUseCase sendSignatureReminderUseCase;

    @MockitoBean
    private SendCancellationConfirmationUseCase sendCancellationConfirmationUseCase;

    @MockitoBean
    private RequestOrderCancellationUseCase requestOrderCancellationUseCase;

    @MockitoBean
    private BookCancellationCostsUseCase bookCancellationCostsUseCase;

    @MockitoBean
    private OrderBikeUseCase orderBikeUseCase;

    @MockitoBean
    private ActivateLeasingUseCase activateLeasingUseCase;

    @BeforeEach
    public void setUp() {
        init(processEngine);
        given(orderBikeUseCase.orderBike(any()))
            .willReturn(new OrderBikeUseCase.Result(new OrderId("ORDER-1"), true));
    }

    @Test
    @DisplayName("happy path - contract signed, bike available, leasing becomes active")
    public void happyPathContractSignedBikeAvailableLeasingBecomesActive() {
        ApplicationId id = submit(35, 3500.0);
        ProcessInstance instance = findProcessInstance(runtimeService, id);

        // validate -> DMN -> conclude-contract sub-process parks on the signature wait state
        continueToNextWaitState(processEngine);

        process.correlateContractSigned(id); // forks into insurance + bike order, joins -> handover wait state
        continueToNextWaitState(processEngine);

        process.correlateHandoverReported(id); // -> withdrawal-period timer
        continueToNextWaitState(processEngine);

        fireTimer(processEngine, Elements.EVENT_WITHDRAWAL_PERIOD_ELAPSED);
        continueToNextWaitState(processEngine);

        assertThat(instance)
            .isEnded()
            .hasPassedInOrder(
                Elements.SERVICE_TASK_VALIDATE_APPLICATION.getValue(),
                Elements.BUSINESS_RULE_TASK_CHECK_CREDIT_RATING.getValue(),
                Elements.SERVICE_TASK_SEND_CONTRACT.getValue(),
                Elements.SERVICE_TASK_ISSUE_INSURANCE_POLICY.getValue(),
                Elements.EVENT_HANDOVER_REPORTED.getValue(),
                Elements.SERVICE_TASK_ACTIVATE_LEASING.getValue(),
                Elements.END_EVENT_LEASING_ACTIVE.getValue())
            .hasNotPassed(
                Elements.END_EVENT_APPLICATION_REJECTED.getValue(),
                Elements.END_EVENT_APPLICATION_CANCELLED.getValue(),
                Elements.END_EVENT_CONTRACT_CANCELLED.getValue());

        then(sendContractUseCase).should(times(1)).sendContract(id);
        then(issueInsurancePolicyUseCase).should(times(1)).issuePolicy(id);
        then(activateLeasingUseCase).should(times(1)).activate(id);
    }

    @Test
    @DisplayName("escalation - contract not signed in time is escalated and rejected")
    public void escalationContractNotSignedInTimeIsEscalatedAndRejected() {
        ApplicationId id = submit(35, 3500.0);
        ProcessInstance instance = findProcessInstance(runtimeService, id);

        continueToNextWaitState(processEngine); // parks on the signature wait state

        fireTimer(processEngine, Elements.EVENT_SIGNATURE_DEADLINE); // deadline -> escalation -> rejection
        continueToNextWaitState(processEngine);

        assertThat(instance)
            .isEnded()
            .hasPassed(
                Elements.EVENT_SIGNATURE_DEADLINE.getValue(),
                Elements.EVENT_CONTRACT_NOT_SIGNED.getValue(),
                Elements.SERVICE_TASK_SEND_REJECTION.getValue(),
                Elements.END_EVENT_APPLICATION_REJECTED.getValue())
            .hasNotPassed(Elements.END_EVENT_LEASING_ACTIVE.getValue());

        then(rejectApplicationUseCase).should(times(1)).reject(id);
    }

    @Test
    @DisplayName("not solvent - the DMN routes the application straight to rejection")
    public void notSolventTheDmnRoutesTheApplicationStraightToRejection() {
        // age below 18 cannot sign a leasing contract, so the DMN returns solvent = false
        ApplicationId id = submit(15, 3500.0);

        continueToNextWaitState(processEngine); // validate -> DMN -> not solvent -> rejection -> end

        then(rejectApplicationUseCase).should(times(1)).reject(id);
        then(sendContractUseCase).should(never()).sendContract(any());
    }

    @Test
    @DisplayName("abort - withdrawing the application compensates the completed steps")
    public void abortWithdrawingTheApplicationCompensatesTheCompletedSteps() {
        given(requestOrderCancellationUseCase.requestCancellation(any())).willReturn(true);

        ApplicationId id = submit(35, 3500.0);
        ProcessInstance instance = findProcessInstance(runtimeService, id);

        // drive up to the handover wait state (contract signed, bike ordered, insured)
        continueToNextWaitState(processEngine);
        process.correlateContractSigned(id);
        continueToNextWaitState(processEngine);

        // Withdrawing triggers compensation; its handlers run in an engine-defined order, so drive the
        // continuations generically until the cancelBikeOrder sub-process parks on its user task.
        process.correlateApplicationWithdrawn(id);
        continueToNextWaitState(processEngine);

        Task task =
            taskService
                .createTaskQuery()
                .taskDefinitionKey(CancelBikeOrderProcessApi.Elements.USER_TASK_CLARIFY_RETURN.getValue())
                .singleResult();
        taskService.complete(task.getId(), Map.of("returnClarified", true));
        continueToNextWaitState(processEngine);

        assertThat(instance)
            .isEnded()
            .hasPassed(
                Elements.SERVICE_TASK_CANCEL_CONTRACT.getValue(),
                Elements.SERVICE_TASK_CANCEL_POLICY.getValue(),
                Elements.CALL_ACTIVITY_CANCEL_BIKE_ORDER.getValue(),
                Elements.SERVICE_TASK_SEND_CANCELLATION_CONFIRMATION.getValue(),
                Elements.END_EVENT_APPLICATION_CANCELLED.getValue())
            .hasNotPassed(Elements.END_EVENT_LEASING_ACTIVE.getValue());

        // External-task delivery is at-least-once: while compensation runs several handlers in
        // parallel, an async continuation firing on the test thread can force an optimistic-locking
        // retry of a handler that already ran, so a worker's use case may be invoked more than once
        // (workers are expected to be idempotent). The process still ends correctly, as asserted above.
        then(cancelContractUseCase).should(atLeast(1)).cancelContract(id);
        then(cancelInsurancePolicyUseCase).should(atLeast(1)).cancelPolicy(id);
        then(sendCancellationConfirmationUseCase).should(atLeast(1)).sendCancellationConfirmation(id);
    }

    @Test
    @DisplayName("bike unavailable - clarifying an alternative re-orders and leasing becomes active")
    public void bikeUnavailableClarifyingAnAlternativeReOrdersAndLeasingBecomesActive() {
        // the first order finds the requested bike unavailable, the re-order after the alternative succeeds
        given(orderBikeUseCase.orderBike(any()))
            .willReturn(
                new OrderBikeUseCase.Result(null, false),
                new OrderBikeUseCase.Result(new OrderId("ORDER-2"), true));

        ApplicationId id = submit(35, 3500.0);
        ProcessInstance instance = findProcessInstance(runtimeService, id);

        continueToNextWaitState(processEngine); // parks on the signature wait state
        process.correlateContractSigned(id);
        continueToNextWaitState(processEngine); // fork -> order finds bike unavailable -> parks on clarify-alternative

        // the alternative is clarified from the outside — the "external" completion of the user task
        process.completeAlternativeClarification(id, true, new BikeId("BIKE-ALT"));
        continueToNextWaitState(processEngine); // re-order succeeds -> parallel join -> handover wait state

        process.correlateHandoverReported(id);
        continueToNextWaitState(processEngine);
        fireTimer(processEngine, Elements.EVENT_WITHDRAWAL_PERIOD_ELAPSED);
        continueToNextWaitState(processEngine);

        assertThat(instance)
            .isEnded()
            .hasPassed(
                Elements.USER_TASK_CLARIFY_ALTERNATIVE.getValue(),
                Elements.SERVICE_TASK_ORDER_BIKE.getValue(),
                Elements.END_EVENT_LEASING_ACTIVE.getValue())
            .hasNotPassed(
                Elements.END_EVENT_CONTRACT_CANCELLED.getValue(),
                Elements.END_EVENT_APPLICATION_REJECTED.getValue());

        then(orderBikeUseCase).should(times(2)).orderBike(id);
    }

    private ApplicationId submit(int age, double income) {
        return submit(age, income, "BIKE-TEST");
    }

    private ApplicationId submit(int age, double income, String bikeId) {
        LeasingApplication application =
            new LeasingApplication(
                ApplicationId.newId(),
                new CustomerName("Test Customer"),
                new Email("test@example.com"),
                age,
                income,
                new BikeId(bikeId),
                LeasingStatus.RECEIVED,
                LocalDateTime.now(),
                null,
                null);
        process.submitRequest(application);
        return application.id();
    }
}

File: service/app/src/test/kotlin/io/miragon/blueprint/process/BikeLeasingProcessTest.kt
package io.miragon.blueprint.process

import com.ninjasquad.springmockk.MockkBean
import io.miragon.blueprint.adapter.process.BikeLeasingProcessProcessApi.Elements
import io.miragon.blueprint.adapter.process.CancelBikeOrderProcessApi
import io.miragon.blueprint.application.port.inbound.ActivateLeasingUseCase
import io.miragon.blueprint.application.port.inbound.BookCancellationCostsUseCase
import io.miragon.blueprint.application.port.inbound.CancelContractUseCase
import io.miragon.blueprint.application.port.inbound.CancelInsurancePolicyUseCase
import io.miragon.blueprint.application.port.inbound.IssueInsurancePolicyUseCase
import io.miragon.blueprint.application.port.inbound.OrderBikeUseCase
import io.miragon.blueprint.application.port.inbound.RejectApplicationUseCase
import io.miragon.blueprint.application.port.inbound.RequestOrderCancellationUseCase
import io.miragon.blueprint.application.port.inbound.SendCancellationConfirmationUseCase
import io.miragon.blueprint.application.port.inbound.SendContractUseCase
import io.miragon.blueprint.application.port.inbound.SendSignatureReminderUseCase
import io.miragon.blueprint.application.port.inbound.ValidateApplicationUseCase
import io.miragon.blueprint.application.port.outbound.LeasingProcess
import io.miragon.blueprint.domain.leasing.ApplicationId
import io.miragon.blueprint.domain.bike.BikeId
import io.miragon.blueprint.domain.leasing.CustomerName
import io.miragon.blueprint.domain.leasing.Email
import io.miragon.blueprint.domain.leasing.LeasingApplication
import io.miragon.blueprint.domain.leasing.LeasingStatus
import io.miragon.blueprint.domain.bike.OrderId
import io.miragon.blueprint.process.util.continueToNextWaitState
import io.miragon.blueprint.process.util.findProcessInstance
import io.miragon.blueprint.process.util.fireTimer
import io.mockk.every
import io.mockk.verify
import org.operaton.bpm.engine.ProcessEngine
import org.operaton.bpm.engine.RuntimeService
import org.operaton.bpm.engine.TaskService
import org.operaton.bpm.engine.test.assertions.bpmn.BpmnAwareTests.assertThat
import org.operaton.bpm.engine.test.assertions.bpmn.BpmnAwareTests.init
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import java.time.LocalDateTime

/**
 * Drives the deployed model end-to-end. Unlike the classic-delegate blueprint, the service tasks are
 * external tasks completed by the real, asynchronously polling `@ProcessEngineWorker` beans (which
 * call the mocked use cases). Each step therefore drives the process with [continueToNextWaitState],
 * which fires async continuations *and* waits for the workers to drain the open external tasks, then
 * correlates messages / fires timers / completes the user task through the [LeasingProcess] port.
 */
@SpringBootTest
@ActiveProfiles("test")
class BikeLeasingProcessTest {

    @Autowired
    private lateinit var process: LeasingProcess

    @Autowired
    private lateinit var runtimeService: RuntimeService

    @Autowired
    private lateinit var taskService: TaskService

    @Autowired
    private lateinit var processEngine: ProcessEngine

    @MockkBean(relaxed = true)
    private lateinit var validateApplicationUseCase: ValidateApplicationUseCase

    @MockkBean(relaxed = true)
    private lateinit var rejectApplicationUseCase: RejectApplicationUseCase

    @MockkBean(relaxed = true)
    private lateinit var sendContractUseCase: SendContractUseCase

    @MockkBean(relaxed = true)
    private lateinit var cancelContractUseCase: CancelContractUseCase

    @MockkBean(relaxed = true)
    private lateinit var issueInsurancePolicyUseCase: IssueInsurancePolicyUseCase

    @MockkBean(relaxed = true)
    private lateinit var cancelInsurancePolicyUseCase: CancelInsurancePolicyUseCase

    @MockkBean(relaxed = true)
    private lateinit var sendSignatureReminderUseCase: SendSignatureReminderUseCase

    @MockkBean(relaxed = true)
    private lateinit var sendCancellationConfirmationUseCase: SendCancellationConfirmationUseCase

    @MockkBean(relaxed = true)
    private lateinit var requestOrderCancellationUseCase: RequestOrderCancellationUseCase

    @MockkBean(relaxed = true)
    private lateinit var bookCancellationCostsUseCase: BookCancellationCostsUseCase

    @MockkBean(relaxed = true)
    private lateinit var orderBikeUseCase: OrderBikeUseCase

    @MockkBean(relaxed = true)
    private lateinit var activateLeasingUseCase: ActivateLeasingUseCase

    @BeforeEach
    fun setUp() {
        init(processEngine)
        every { orderBikeUseCase.orderBike(any()) } returns
            OrderBikeUseCase.Result(OrderId("ORDER-1"), bikeAvailable = true)
    }

    @Test
    fun `happy path - contract signed, bike available, leasing becomes active`() {
        val id = submit(age = 35, income = 3500.0)
        val instance = runtimeService.findProcessInstance(id)

        // validate -> DMN -> conclude-contract sub-process parks on the signature wait state
        processEngine.continueToNextWaitState()

        process.correlateContractSigned(id) // forks into insurance + bike order, joins -> handover wait state
        processEngine.continueToNextWaitState()

        process.correlateHandoverReported(id) // -> withdrawal-period timer
        processEngine.continueToNextWaitState()

        processEngine.fireTimer(Elements.EVENT_WITHDRAWAL_PERIOD_ELAPSED)
        processEngine.continueToNextWaitState()

        assertThat(instance)
            .isEnded
            .hasPassedInOrder(
                Elements.SERVICE_TASK_VALIDATE_APPLICATION.value,
                Elements.BUSINESS_RULE_TASK_CHECK_CREDIT_RATING.value,
                Elements.SERVICE_TASK_SEND_CONTRACT.value,
                Elements.SERVICE_TASK_ISSUE_INSURANCE_POLICY.value,
                Elements.EVENT_HANDOVER_REPORTED.value,
                Elements.SERVICE_TASK_ACTIVATE_LEASING.value,
                Elements.END_EVENT_LEASING_ACTIVE.value,
            )
            .hasNotPassed(
                Elements.END_EVENT_APPLICATION_REJECTED.value,
                Elements.END_EVENT_APPLICATION_CANCELLED.value,
                Elements.END_EVENT_CONTRACT_CANCELLED.value,
            )

        verify(exactly = 1) { sendContractUseCase.sendContract(id) }
        verify(exactly = 1) { issueInsurancePolicyUseCase.issuePolicy(id) }
        verify(exactly = 1) { activateLeasingUseCase.activate(id) }
    }

    @Test
    fun `escalation - contract not signed in time is escalated and rejected`() {
        val id = submit(age = 35, income = 3500.0)
        val instance = runtimeService.findProcessInstance(id)

        processEngine.continueToNextWaitState() // parks on the signature wait state

        processEngine.fireTimer(Elements.EVENT_SIGNATURE_DEADLINE) // deadline -> escalation -> rejection
        processEngine.continueToNextWaitState()

        assertThat(instance)
            .isEnded
            .hasPassed(
                Elements.EVENT_SIGNATURE_DEADLINE.value,
                Elements.EVENT_CONTRACT_NOT_SIGNED.value,
                Elements.SERVICE_TASK_SEND_REJECTION.value,
                Elements.END_EVENT_APPLICATION_REJECTED.value,
            )
            .hasNotPassed(Elements.END_EVENT_LEASING_ACTIVE.value)

        verify(exactly = 1) { rejectApplicationUseCase.reject(id) }
    }

    @Test
    fun `not solvent - the DMN routes the application straight to rejection`() {
        // age below 18 cannot sign a leasing contract, so the DMN returns solvent = false
        val id = submit(age = 15, income = 3500.0)

        processEngine.continueToNextWaitState() // validate -> DMN -> not solvent -> rejection -> end

        verify(exactly = 1) { rejectApplicationUseCase.reject(id) }
        verify(exactly = 0) { sendContractUseCase.sendContract(any()) }
    }

    @Test
    fun `abort - withdrawing the application compensates the completed steps`() {
        every { requestOrderCancellationUseCase.requestCancellation(any()) } returns true

        val id = submit(age = 35, income = 3500.0)
        val instance = runtimeService.findProcessInstance(id)

        // drive up to the handover wait state (contract signed, bike ordered, insured)
        processEngine.continueToNextWaitState()
        process.correlateContractSigned(id)
        processEngine.continueToNextWaitState()

        // Withdrawing triggers compensation; its handlers run in an engine-defined order, so drive the
        // continuations generically until the cancelBikeOrder sub-process parks on its user task.
        process.correlateApplicationWithdrawn(id)
        processEngine.continueToNextWaitState()

        val task =
            taskService
                .createTaskQuery()
                .taskDefinitionKey(CancelBikeOrderProcessApi.Elements.USER_TASK_CLARIFY_RETURN.value)
                .singleResult()
        taskService.complete(task.id, mapOf("returnClarified" to true))
        processEngine.continueToNextWaitState()

        assertThat(instance)
            .isEnded
            .hasPassed(
                Elements.SERVICE_TASK_CANCEL_CONTRACT.value,
                Elements.SERVICE_TASK_CANCEL_POLICY.value,
                Elements.CALL_ACTIVITY_CANCEL_BIKE_ORDER.value,
                Elements.SERVICE_TASK_SEND_CANCELLATION_CONFIRMATION.value,
                Elements.END_EVENT_APPLICATION_CANCELLED.value,
            )
            .hasNotPassed(Elements.END_EVENT_LEASING_ACTIVE.value)

        // External-task delivery is at-least-once: while compensation runs several handlers in
        // parallel, an async continuation firing on the test thread can force an optimistic-locking
        // retry of a handler that already ran, so a worker's use case may be invoked more than once
        // (workers are expected to be idempotent). The process still ends correctly, as asserted above.
        verify(atLeast = 1) { cancelContractUseCase.cancelContract(id) }
        verify(atLeast = 1) { cancelInsurancePolicyUseCase.cancelPolicy(id) }
        verify(atLeast = 1) { sendCancellationConfirmationUseCase.sendCancellationConfirmation(id) }
    }

    @Test
    fun `bike unavailable - clarifying an alternative re-orders and leasing becomes active`() {
        // the first order finds the requested bike unavailable, the re-order after the alternative succeeds
        every { orderBikeUseCase.orderBike(any()) } returnsMany
            listOf(
                OrderBikeUseCase.Result(orderId = null, bikeAvailable = false),
                OrderBikeUseCase.Result(OrderId("ORDER-2"), bikeAvailable = true),
            )

        val id = submit(age = 35, income = 3500.0)
        val instance = runtimeService.findProcessInstance(id)

        processEngine.continueToNextWaitState() // parks on the signature wait state
        process.correlateContractSigned(id)
        processEngine.continueToNextWaitState() // fork -> order finds bike unavailable -> parks on clarify-alternative

        // the alternative is clarified from the outside — the "external" completion of the user task
        process.completeAlternativeClarification(id, alternativeFound = true, bikeId = BikeId("BIKE-ALT"))
        processEngine.continueToNextWaitState() // re-order succeeds -> parallel join -> handover wait state

        process.correlateHandoverReported(id)
        processEngine.continueToNextWaitState()
        processEngine.fireTimer(Elements.EVENT_WITHDRAWAL_PERIOD_ELAPSED)
        processEngine.continueToNextWaitState()

        assertThat(instance)
            .isEnded
            .hasPassed(
                Elements.USER_TASK_CLARIFY_ALTERNATIVE.value,
                Elements.SERVICE_TASK_ORDER_BIKE.value,
                Elements.END_EVENT_LEASING_ACTIVE.value,
            )
            .hasNotPassed(
                Elements.END_EVENT_CONTRACT_CANCELLED.value,
                Elements.END_EVENT_APPLICATION_REJECTED.value,
            )

        verify(exactly = 2) { orderBikeUseCase.orderBike(id) }
    }

    private fun submit(age: Int, income: Double, bikeId: String = "BIKE-TEST"): ApplicationId {
        val application =
            LeasingApplication(
                id = ApplicationId.new(),
                customerName = CustomerName("Test Customer"),
                email = Email("test@example.com"),
                age = age,
                monthlyNetIncome = income,
                bikeId = BikeId(bikeId),
                status = LeasingStatus.RECEIVED,
                createdAt = LocalDateTime.now(),
            )
        process.submitRequest(application)
        return application.id
    }
}
